File: scripts/vlm_stress_test.py
#!/usr/bin/env python3
"""
VLM Stress Test — Vigie Bot (Python version, aligned with project architecture).

Runs 100 VLM calls on a fixed screenshot, then repeats. Measures latency
(p50/p95/p99), success rate, JSON parse rate, and response stability.

Uses the `z-ai vision` CLI in subprocess (same pattern as app/services/zai_llm.py).

Outputs:
  /home/z/my-project/scripts/vlm_stress_run1.json
  /home/z/my-project/scripts/vlm_stress_run2.json
  /home/z/my-project/scripts/vlm_stress_report.md
"""

from __future__ import annotations

import asyncio
import hashlib
import json
import os
import statistics
import time
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

# ---------- Config ----------
SCREENSHOT_PATH = "/home/z/my-project/scripts/vigie_screenshot.png"
NUM_CALLS = int(os.environ.get("VLM_STRESS_NUM_CALLS", "100"))
RUNS = int(os.environ.get("VLM_STRESS_RUNS", "2"))
OUTPUT_DIR = Path("/home/z/my-project/scripts")
REPORT_PATH = OUTPUT_DIR / "vlm_stress_report.md"
CONCURRENCY = 1  # sequential — VLM API rate-limits aggressively
INTER_CALL_DELAY = 0.5  # seconds between successful calls
MAX_RETRIES = 3  # retry on 429 with exponential backoff
RETRY_BASE_DELAY = 10.0  # seconds (10s, 20s, 40s)

SYSTEM_PROMPT = """Tu es Vigie-VLM, un assistant visuel dédié à l'analyse de captures d'écran du bot Slack Vigie (prévention des décès par isolation pendant les canicules).

Ta mission : analyser une capture d'écran et produire un rapport structuré contenant :
1. COVERAGE_PERCENT: pourcentage de couverture affiché (ou "N/A")
2. AVG_LATENCY_MIN: latence moyenne de check-in affichée en minutes (ou "N/A")
3. L2_COUNT: nombre d'escalades L2 (pas de réponse)
4. L3_COUNT: nombre d'escalades L3 (inconscient)
5. CRISIS_MSG_COUNT: nombre de messages dans #cellule-crise
6. TOP_SECTORS: liste des secteurs visibles (JSON array)
7. ACTIVE_ALERTS: bénéficiaires actuellement en alerte (JSON array d'objets {name, level})
8. DASHBOARD_HEALTH: "OK" si coverage >= 90% ET L3=0, sinon "ALERT"
9. SUMMARY: 1 phrase en français résumant l'état global

Réponds UNIQUEMENT en JSON valide, sans texte autour."""

# z-ai vision CLI has no --system flag, so we concatenate the system prompt
# with the user prompt into a single --prompt argument.
USER_PROMPT = SYSTEM_PROMPT + "\n\n" + "Analyse cette capture du dashboard Vigie et réponds en JSON."


# ---------- Helpers ----------
def percentile(sorted_list: list[float], p: float) -> float:
    if not sorted_list:
        return 0.0
    idx = min(int((p / 100.0) * len(sorted_list)), len(sorted_list) - 1)
    return sorted_list[idx]


def hash_str(s: str) -> str:
    return hashlib.md5(s.encode("utf-8")).hexdigest()


async def call_vlm_once(sem: asyncio.Semaphore, idx: int) -> dict[str, Any]:
    """One VLM call via z-ai CLI subprocess, with retry on 429."""
    async with sem:
        start = time.perf_counter()
        last_error = ""
        for attempt in range(MAX_RETRIES + 1):
            try:
                proc = await asyncio.create_subprocess_exec(
                    "z-ai", "vision",
                    "-p", USER_PROMPT,
                    "-i", SCREENSHOT_PATH,
                    stdout=asyncio.subprocess.PIPE,
                    stderr=asyncio.subprocess.PIPE,
                )
                stdout, stderr = await asyncio.wait_for(proc.communicate(), timeout=180.0)

                if proc.returncode != 0:
                    err_text = stderr.decode("utf-8", errors="replace")
                    last_error = f"rc={proc.returncode} stderr={err_text[:300]}"
                    # Check for 429 rate limit
                    if "429" in err_text and attempt < MAX_RETRIES:
                        backoff = RETRY_BASE_DELAY * (2 ** attempt)
                        print(f"    [idx={idx}] 429 hit, backing off {backoff:.0f}s (attempt {attempt+1}/{MAX_RETRIES})", flush=True)
                        await asyncio.sleep(backoff)
                        continue
                    return {
                        "index": idx,
                        "ok": False,
                        "latencyMs": (time.perf_counter() - start) * 1000.0,
                        "error": last_error,
                    }

                # z-ai vision prints emojis/log lines, then a JSON block.
                raw = stdout.decode("utf-8", errors="replace")
                content = ""
                try:
                    start_json = raw.find("{")
                    if start_json == -1:
                        raise ValueError("no JSON found in stdout")
                    decoder = json.JSONDecoder()
                    data, _ = decoder.raw_decode(raw[start_json:])
                    if isinstance(data, dict):
                        choices = data.get("choices") or []
                        if choices:
                            msg = choices[0].get("message", {})
                            content = msg.get("content", "") or ""
                        if not content:
                            content = data.get("content", "") or data.get("response", "") or raw
                    else:
                        content = str(data)
                except (json.JSONDecodeError, ValueError):
                    content = raw.strip()

                parsed = None
                try:
                    import re
                    m = re.search(r"\{[\s\S]*\}", content)
                    if m:
                        parsed = json.loads(m.group(0))
                except Exception:
                    parsed = None

                latency_ms = (time.perf_counter() - start) * 1000.0
                # Inter-call delay to avoid hammering the API
                await asyncio.sleep(INTER_CALL_DELAY)

                return {
                    "index": idx,
                    "ok": True,
                    "latencyMs": latency_ms,
                    "content": content,
                    "parsed": parsed,
                    "retries": attempt,
                }
            except asyncio.TimeoutError:
                last_error = "timeout_180s"
                if attempt < MAX_RETRIES:
                    print(f"    [idx={idx}] timeout, retrying (attempt {attempt+1}/{MAX_RETRIES})", flush=True)
                    await asyncio.sleep(RETRY_BASE_DELAY)
                    continue
                return {
                    "index": idx,
                    "ok": False,
                    "latencyMs": (time.perf_counter() - start) * 1000.0,
                    "error": last_error,
                }
            except Exception as e:
                last_error = f"exc={type(e).__name__}:{str(e)[:200]}"
                if attempt < MAX_RETRIES:
                    await asyncio.sleep(RETRY_BASE_DELAY)
                    continue
                return {
                    "index": idx,
                    "ok": False,
                    "latencyMs": (time.perf_counter() - start) * 1000.0,
                    "error": last_error,
                }

        # Should not reach here, but safety net
        return {
            "index": idx,
            "ok": False,
            "latencyMs": (time.perf_counter() - start) * 1000.0,
            "error": f"exhausted_retries: {last_error}",
        }


async def run_pass(run_num: int) -> dict[str, Any]:
    started_at = datetime.now(timezone.utc).isoformat()
    start_total = time.perf_counter()

    print(f"\n=== Run {run_num} — {NUM_CALLS} VLM calls (concurrency={CONCURRENCY}) ===", flush=True)

    sem = asyncio.Semaphore(CONCURRENCY)
    tasks = [call_vlm_once(sem, i) for i in range(NUM_CALLS)]

    results: list[dict[str, Any]] = []
    # Stream completion with progress every 10
    done = 0
    for coro in asyncio.as_completed(tasks):
        r = await coro
        results.append(r)
        done += 1
        if done % 10 == 0 or done == NUM_CALLS:
            ok = sum(1 for r in results if r.get("ok"))
            last = results[-1]
            print(
                f"  [{done}/{NUM_CALLS}] ok={ok} last={last.get('latencyMs', 0):.0f}ms"
                + ("" if last.get("ok") else f" ERR={(last.get('error') or '')[:80]}"),
                flush=True,
            )

    # Order by index for reproducibility
    results.sort(key=lambda r: r.get("index", 0))

    total_ms = (time.perf_counter() - start_total) * 1000.0
    ended_at = datetime.now(timezone.utc).isoformat()

    ok_results = [r for r in results if r.get("ok")]
    latencies = sorted([float(r["latencyMs"]) for r in ok_results])
    parse_count = sum(1 for r in ok_results if r.get("parsed") is not None)

    response_hash_counts: dict[str, int] = {}
    for r in ok_results:
        h = hash_str(r.get("content", ""))
        response_hash_counts[h] = response_hash_counts.get(h, 0) + 1

    stats = {
        "successRate": len(ok_results) / max(1, len(results)),
        "p50Ms": percentile(latencies, 50),
        "p95Ms": percentile(latencies, 95),
        "p99Ms": percentile(latencies, 99),
        "avgMs": statistics.mean(latencies) if latencies else 0.0,
        "minMs": latencies[0] if latencies else 0.0,
        "maxMs": latencies[-1] if latencies else 0.0,
        "parseRate": parse_count / max(1, len(ok_results)),
        "uniqueResponses": len(response_hash_counts),
        "responseHashCounts": response_hash_counts,
    }

    print(
        f"  -> Run {run_num} done in {total_ms/1000:.1f}s "
        f"success={stats['successRate']:.2%} p50={stats['p50Ms']:.0f}ms "
        f"p95={stats['p95Ms']:.0f}ms parse={stats['parseRate']:.2%} "
        f"unique={stats['uniqueResponses']}/{len(ok_results)}",
        flush=True,
    )

    return {
        "run": run_num,
        "startedAt": started_at,
        "endedAt": ended_at,
        "totalMs": total_ms,
        "results": results,
        "stats": stats,
    }


def build_report(runs: list[dict[str, Any]]) -> str:
    r1, r2 = runs[0], runs[1]
    ok1 = sum(1 for r in r1["results"] if r.get("ok"))
    ok2 = sum(1 for r in r2["results"] if r.get("ok"))
    top1 = sorted(r1["stats"]["responseHashCounts"].items(), key=lambda kv: -kv[1])[0] if r1["stats"]["responseHashCounts"] else None
    top2 = sorted(r2["stats"]["responseHashCounts"].items(), key=lambda kv: -kv[1])[0] if r2["stats"]["responseHashCounts"] else None

    sample1 = next((r.get("parsed") for r in r1["results"] if r.get("parsed")), None)
    sample2 = next((r.get("parsed") for r in r2["results"] if r.get("parsed")), None)

    latency_drift = (
        (r2["stats"]["avgMs"] - r1["stats"]["avgMs"]) / r1["stats"]["avgMs"] * 100
        if r1["stats"]["avgMs"] else 0.0
    )

    stability1 = (top1[1] / ok1 * 100) if top1 and ok1 else 0
    stability2 = (top2[1] / ok2 * 100) if top2 and ok2 else 0

    md = f"""# VLM Stress Test Report — Vigie Bot

**Date:** {datetime.now(timezone.utc).isoformat()}
**Screenshot:** `{SCREENSHOT_PATH}`
**Configuration:** {NUM_CALLS} calls x {RUNS} runs = {NUM_CALLS * RUNS} total VLM calls
**Concurrency:** {CONCURRENCY} parallel subprocess calls
**Backend:** `z-ai vision` CLI (subprocess)

## 1. Summary

| Metric | Run 1 | Run 2 | Drift |
|---|---|---|---|
| Success rate | {r1['stats']['successRate']*100:.1f}% | {r2['stats']['successRate']*100:.1f}% | {(r2['stats']['successRate']-r1['stats']['successRate'])*100:+.1f} pts |
| Parse rate (JSON valid) | {r1['stats']['parseRate']*100:.1f}% | {r2['stats']['parseRate']*100:.1f}% | {(r2['stats']['parseRate']-r1['stats']['parseRate'])*100:+.1f} pts |
| Avg latency | {r1['stats']['avgMs']:.0f} ms | {r2['stats']['avgMs']:.0f} ms | {latency_drift:+.1f}% |
| p50 latency | {r1['stats']['p50Ms']:.0f} ms | {r2['stats']['p50Ms']:.0f} ms | - |
| p95 latency | {r1['stats']['p95Ms']:.0f} ms | {r2['stats']['p95Ms']:.0f} ms | - |
| p99 latency | {r1['stats']['p99Ms']:.0f} ms | {r2['stats']['p99Ms']:.0f} ms | - |
| Min latency | {r1['stats']['minMs']:.0f} ms | {r2['stats']['minMs']:.0f} ms | - |
| Max latency | {r1['stats']['maxMs']:.0f} ms | {r2['stats']['maxMs']:.0f} ms | - |
| Total wall clock | {r1['totalMs']/1000:.1f} s | {r2['totalMs']/1000:.1f} s | - |
| Unique response hashes | {r1['stats']['uniqueResponses']} | {r2['stats']['uniqueResponses']} | - |
| Top hash frequency | {top1[1] if top1 else 0}/{ok1} | {top2[1] if top2 else 0}/{ok2} | - |

## 2. Stability Analysis

- **Run 1 stability:** {stability1:.1f}% of responses were identical (top hash)
- **Run 2 stability:** {stability2:.1f}% of responses were identical (top hash)
- **Cross-run latency drift:** {latency_drift:+.1f}% ({"HIGH" if abs(latency_drift) > 20 else "MODERATE" if abs(latency_drift) > 10 else "LOW"})
- **Cross-run stability drift:** {abs(stability2 - stability1):.1f} pts

## 3. Sample Parsed Response (Run 1)

```json
{json.dumps(sample1 or {"error": "no parsed sample"}, indent=2, ensure_ascii=False)}
```

## 4. Sample Parsed Response (Run 2)

```json
{json.dumps(sample2 or {"error": "no parsed sample"}, indent=2, ensure_ascii=False)}
```

## 5. Errors (first 5)

"""
    errors = [r for r in [*r1["results"], *r2["results"]] if not r.get("ok")][:5]
    if not errors:
        md += f"No errors across {NUM_CALLS * RUNS} calls.\n\n"
    else:
        md += "| Run | Index | Latency | Error |\n|---|---|---|---|\n"
        for e in errors:
            run_num = 1 if e.get("index", 0) < NUM_CALLS else 2
            md += f"| {run_num} | {e.get('index', 0) % NUM_CALLS} | {e.get('latencyMs', 0):.0f}ms | {(e.get('error') or '')[:100].replace('|', '\\|')} |\n"
        md += "\n"

    md += """## 6. Verdict

"""
    verdict = (
        "PASS — VLM is stable enough for production integration in Vigie bot."
        if (
            r1["stats"]["successRate"] >= 0.95
            and r2["stats"]["successRate"] >= 0.95
            and r1["stats"]["parseRate"] >= 0.9
            and r2["stats"]["parseRate"] >= 0.9
            and abs(latency_drift) < 25
        )
        else "CONDITIONAL — review metrics before production integration."
    )
    md += f"**{verdict}**\n"
    return md


async def main() -> None:
    print("VLM Stress Test — Vigie Bot (Python)")
    print(f"Screenshot: {SCREENSHOT_PATH}")
    print(f"Config: {NUM_CALLS} calls x {RUNS} runs, concurrency={CONCURRENCY}")

    if not Path(SCREENSHOT_PATH).exists():
        raise SystemExit(f"Screenshot not found: {SCREENSHOT_PATH}")

    runs: list[dict[str, Any]] = []
    for i in range(1, RUNS + 1):
        result = await run_pass(i)
        runs.append(result)
        out_json = OUTPUT_DIR / f"vlm_stress_run{i}.json"
        out_json.write_text(json.dumps(result, indent=2, ensure_ascii=False))
        print(f"  -> Saved run {i} to {out_json.name}", flush=True)
        if i < RUNS:
            print("  -> Cooling down 5s before next run...", flush=True)
            await asyncio.sleep(5)

    report = build_report(runs)
    REPORT_PATH.write_text(report)
    print(f"\nReport saved to: {REPORT_PATH}")
    print("\n--- REPORT PREVIEW ---\n")
    print(report[:2500])


if __name__ == "__main__":
    asyncio.run(main())

File: scripts/vlm_stress_test.ts
/**
 * VLM Stress Test — Vigie Bot
 *
 * Run 100 VLM calls on a fixed screenshot, then repeat.
 * Measure: latency (p50/p95/p99), token usage, response stability.
 *
 * Usage:
 *   npx tsx /home/z/my-project/scripts/vlm_stress_test.ts
 *
 * Outputs:
 *   /home/z/my-project/scripts/vlm_stress_run1.json
 *   /home/z/my-project/scripts/vlm_stress_run2.json
 *   /home/z/my-project/scripts/vlm_stress_report.md
 */

import ZAI from 'z-ai-web-dev-sdk';
import * as fs from 'fs';
import * as path from 'path';

// ---------- Config ----------
const SCREENSHOT_PATH = '/home/z/my-project/scripts/vigie_screenshot.png';
const NUM_CALLS = 100;
const RUNS = 2;
const OUTPUT_DIR = '/home/z/my-project/scripts';
const REPORT_PATH = path.join(OUTPUT_DIR, 'vlm_stress_report.md');

const SYSTEM_PROMPT = `Tu es Vigie-VLM, un assistant visuel dédié à l'analyse de captures d'écran du bot Slack Vigie (prévention des décès par isolation pendant les canicules).

Ta mission : analyser une capture d'écran et produire un rapport structuré contenant :
1. COVERAGE_PERCENT: pourcentage de couverture affiché (ou "N/A")
2. AVG_LATENCY_MIN: latence moyenne de check-in affichée en minutes (ou "N/A")
3. L2_COUNT: nombre d'escalades L2 (pas de réponse)
4. L3_COUNT: nombre d'escalades L3 (inconscient)
5. CRISIS_MSG_COUNT: nombre de messages dans #cellule-crise
6. TOP_SECTORS: liste des secteurs visibles (JSON array)
7. ACTIVE_ALERTS: bénéficiaires actuellement en alerte (JSON array d'objets {name, level})
8. DASHBOARD_HEALTH: "OK" si coverage >= 90% ET L3=0, sinon "ALERT"
9. SUMMARY: 1 phrase en français résumant l'état global

Réponds UNIQUEMENT en JSON valide, sans texte autour.`;

interface CallResult {
  index: number;
  ok: boolean;
  latencyMs: number;
  content?: string;
  parsed?: any;
  error?: string;
}

interface RunResult {
  run: number;
  startedAt: string;
  endedAt: string;
  totalMs: number;
  results: CallResult[];
  stats: {
    successRate: number;
    p50Ms: number;
    p95Ms: number;
    p99Ms: number;
    avgMs: number;
    minMs: number;
    maxMs: number;
    parseRate: number;
    uniqueResponses: number;
    responseHashCounts: Record<string, number>;
  };
}

// ---------- Helpers ----------
function loadScreenshotAsBase64(): string {
  if (!fs.existsSync(SCREENSHOT_PATH)) {
    throw new Error(`Screenshot not found: ${SCREENSHOT_PATH}`);
  }
  const buf = fs.readFileSync(SCREENSHOT_PATH);
  return `data:image/png;base64,${buf.toString('base64')}`;
}

function percentile(sorted: number[], p: number): number {
  if (sorted.length === 0) return 0;
  const idx = Math.min(Math.floor((p / 100) * sorted.length), sorted.length - 1);
  return sorted[idx];
}

function hashStr(s: string): string {
  let h = 5381;
  for (let i = 0; i < s.length; i++) {
    h = (h * 33) ^ s.charCodeAt(i);
  }
  return (h >>> 0).toString(16);
}

async function callVlm(zai: any, b64: string): Promise<CallResult> {
  const start = Date.now();
  try {
    const response = await zai.chat.completions.createVision({
      messages: [
        {
          role: 'system',
          content: [{ type: 'text', text: SYSTEM_PROMPT }]
        },
        {
          role: 'user',
          content: [
            { type: 'text', text: 'Analyse cette capture du dashboard Vigie et réponds en JSON.' },
            { type: 'image_url', image_url: { url: b64 } }
          ]
        }
      ],
      thinking: { type: 'disabled' }
    });
    const latencyMs = Date.now() - start;
    const content = response.choices?.[0]?.message?.content ?? '';
    let parsed: any = undefined;
    try {
      // Try to extract JSON from possibly-wrapped response
      const jsonMatch = content.match(/\{[\s\S]*\}/);
      parsed = jsonMatch ? JSON.parse(jsonMatch[0]) : undefined;
    } catch {
      parsed = undefined;
    }
    return { index: 0, ok: true, latencyMs, content, parsed };
  } catch (err: any) {
    return {
      index: 0,
      ok: false,
      latencyMs: Date.now() - start,
      error: err?.message ?? String(err)
    };
  }
}

async function runPass(zai: any, b64: string, runNum: number): Promise<RunResult> {
  const startedAt = new Date().toISOString();
  const startTotal = Date.now();
  const results: CallResult[] = [];

  console.log(`\n=== Run ${runNum} — ${NUM_CALLS} VLM calls ===`);

  for (let i = 0; i < NUM_CALLS; i++) {
    const r = await callVlm(zai, b64);
    r.index = i;
    results.push(r);

    if (i % 10 === 0 || i === NUM_CALLS - 1) {
      const ok = results.filter(r => r.ok).length;
      const last = results[results.length - 1];
      console.log(
        `  [${i + 1}/${NUM_CALLS}] ok=${ok} last=${last.latencyMs}ms ${
          last.ok ? '' : 'ERR=' + (last.error ?? '').slice(0, 80)
        }`
      );
    }
  }

  const totalMs = Date.now() - startTotal;
  const endedAt = new Date().toISOString();

  const okResults = results.filter(r => r.ok);
  const latencies = okResults.map(r => r.latencyMs).sort((a, b) => a - b);
  const parseCount = okResults.filter(r => r.parsed !== undefined).length;
  const responseHashCounts: Record<string, number> = {};
  for (const r of okResults) {
    const h = hashStr(r.content ?? '');
    responseHashCounts[h] = (responseHashCounts[h] ?? 0) + 1;
  }

  const stats = {
    successRate: okResults.length / results.length,
    p50Ms: percentile(latencies, 50),
    p95Ms: percentile(latencies, 95),
    p99Ms: percentile(latencies, 99),
    avgMs: latencies.reduce((a, b) => a + b, 0) / Math.max(1, latencies.length),
    minMs: latencies[0] ?? 0,
    maxMs: latencies[latencies.length - 1] ?? 0,
    parseRate: parseCount / Math.max(1, okResults.length),
    uniqueResponses: Object.keys(responseHashCounts).length,
    responseHashCounts
  };

  console.log(`  -> Run ${runNum} done in ${(totalMs / 1000).toFixed(1)}s`);
  console.log(`     success=${stats.successRate} p50=${stats.p50Ms}ms p95=${stats.p95Ms}ms parse=${stats.parseRate}`);
  console.log(`     unique responses=${stats.uniqueResponses} (out of ${okResults.length})`);

  return { run: runNum, startedAt, endedAt, totalMs, results, stats };
}

function buildReport(runs: RunResult[]): string {
  const r1 = runs[0];
  const r2 = runs[1];

  const topHash1 = Object.entries(r1.stats.responseHashCounts).sort((a, b) => b[1] - a[1])[0];
  const topHash2 = Object.entries(r2.stats.responseHashCounts).sort((a, b) => b[1] - a[1])[0];

  // Sample a parsed response from each run
  const sample1 = r1.results.find(r => r.parsed)?.parsed;
  const sample2 = r2.results.find(r => r.parsed)?.parsed;

  const latencyDrift = ((r2.stats.avgMs - r1.stats.avgMs) / r1.stats.avgMs) * 100;

  let md = `# VLM Stress Test Report — Vigie Bot\n\n`;
  md += `**Date:** ${new Date().toISOString()}\n`;
  md += `**Screenshot:** \`${SCREENSHOT_PATH}\`\n`;
  md += `**Configuration:** ${NUM_CALLS} calls × ${RUNS} runs = ${NUM_CALLS * RUNS} total VLM calls\n\n`;

  md += `## 1. Summary\n\n`;
  md += `| Metric | Run 1 | Run 2 | Drift |\n`;
  md += `|---|---|---|---|\n`;
  md += `| Success rate | ${(r1.stats.successRate * 100).toFixed(1)}% | ${(r2.stats.successRate * 100).toFixed(1)}% | ${((r2.stats.successRate - r1.stats.successRate) * 100).toFixed(1)} pts |\n`;
  md += `| Parse rate (JSON valid) | ${(r1.stats.parseRate * 100).toFixed(1)}% | ${(r2.stats.parseRate * 100).toFixed(1)}% | ${((r2.stats.parseRate - r1.stats.parseRate) * 100).toFixed(1)} pts |\n`;
  md += `| Avg latency | ${r1.stats.avgMs.toFixed(0)} ms | ${r2.stats.avgMs.toFixed(0)} ms | ${latencyDrift.toFixed(1)}% |\n`;
  md += `| p50 latency | ${r1.stats.p50Ms} ms | ${r2.stats.p50Ms} ms | - |\n`;
  md += `| p95 latency | ${r1.stats.p95Ms} ms | ${r2.stats.p95Ms} ms | - |\n`;
  md += `| p99 latency | ${r1.stats.p99Ms} ms | ${r2.stats.p99Ms} ms | - |\n`;
  md += `| Min latency | ${r1.stats.minMs} ms | ${r2.stats.minMs} ms | - |\n`;
  md += `| Max latency | ${r1.stats.maxMs} ms | ${r2.stats.maxMs} ms | - |\n`;
  md += `| Total wall clock | ${(r1.totalMs / 1000).toFixed(1)} s | ${(r2.totalMs / 1000).toFixed(1)} s | - |\n`;
  md += `| Unique response hashes | ${r1.stats.uniqueResponses} | ${r2.stats.uniqueResponses} | - |\n`;
  md += `| Top hash frequency | ${topHash1 ? topHash1[1] : 0}/${r1.results.filter(r => r.ok).length} | ${topHash2 ? topHash2[1] : 0}/${r2.results.filter(r => r.ok).length} | - |\n\n`;

  md += `## 2. Stability Analysis\n\n`;
  const stability1 = topHash1 ? (topHash1[1] / r1.results.filter(r => r.ok).length) * 100 : 0;
  const stability2 = topHash2 ? (topHash2[1] / r2.results.filter(r => r.ok).length) * 100 : 0;
  md += `- **Run 1 stability:** ${stability1.toFixed(1)}% of responses were identical (top hash)\n`;
  md += `- **Run 2 stability:** ${stability2.toFixed(1)}% of responses were identical (top hash)\n`;
  md += `- **Cross-run drift:** ${latencyDrift > 20 ? 'HIGH' : latencyDrift > 10 ? 'MODERATE' : 'LOW'} latency drift (${latencyDrift.toFixed(1)}%)\n\n`;

  md += `## 3. Sample Parsed Response (Run 1)\n\n`;
  md += '```json\n' + JSON.stringify(sample1 ?? { error: 'no parsed sample' }, null, 2) + '\n```\n\n';

  md += `## 4. Sample Parsed Response (Run 2)\n\n`;
  md += '```json\n' + JSON.stringify(sample2 ?? { error: 'no parsed sample' }, null, 2) + '\n```\n\n';

  md += `## 5. Errors (first 5)\n\n`;
  const errors = [...r1.results, ...r2.results].filter(r => !r.ok).slice(0, 5);
  if (errors.length === 0) {
    md += `No errors across ${NUM_CALLS * RUNS} calls.\n\n`;
  } else {
    md += `| Run | Index | Latency | Error |\n|---|---|---|---|\n`;
    for (const e of errors) {
      md += `| ${e.index < NUM_CALLS ? 1 : 2} | ${e.index % NUM_CALLS} | ${e.latencyMs}ms | ${(e.error ?? '').slice(0, 100).replace(/\|/g, '\\|')} |\n`;
    }
    md += '\n';
  }

  md += `## 6. Verdict\n\n`;
  const verdict =
    r1.stats.successRate >= 0.95 &&
    r2.stats.successRate >= 0.95 &&
    r1.stats.parseRate >= 0.9 &&
    r2.stats.parseRate >= 0.9 &&
    Math.abs(latencyDrift) < 25
      ? 'PASS — VLM is stable enough for production integration in Vigie bot.'
      : 'CONDITIONAL — review metrics before production integration.';
  md += `**${verdict}**\n`;

  return md;
}

// ---------- Main ----------
async function main() {
  console.log('VLM Stress Test — Vigie Bot');
  console.log(`Screenshot: ${SCREENSHOT_PATH}`);
  console.log(`Config: ${NUM_CALLS} calls × ${RUNS} runs`);

  const b64 = loadScreenshotAsBase64();
  console.log(`Screenshot base64 length: ${b64.length} chars`);

  const zai = await ZAI.create();

  const runs: RunResult[] = [];
  for (let i = 1; i <= RUNS; i++) {
    const result = await runPass(zai, b64, i);
    runs.push(result);
    fs.writeFileSync(
      path.join(OUTPUT_DIR, `vlm_stress_run${i}.json`),
      JSON.stringify(result, null, 2)
    );
    console.log(`  -> Saved run ${i} to vlm_stress_run${i}.json`);

    if (i < RUNS) {
      console.log('  -> Cooling down 5s before next run...');
      await new Promise(r => setTimeout(r, 5000));
    }
  }

  const report = buildReport(runs);
  fs.writeFileSync(REPORT_PATH, report);
  console.log(`\nReport saved to: ${REPORT_PATH}`);
  console.log('\n--- REPORT PREVIEW ---\n');
  console.log(report.slice(0, 2000));
}

main().catch(err => {
  console.error('FATAL:', err);
  process.exit(1);
});
